### ISSUE TITLE
storage: abstract disk I/O

### ISSUE THREAD

**Issue by skyzh** [risinglightdb/risinglight#660]
Currently, we have create_dir, etc. everywhere. We'd better have a single interface to operate files on disk, so as to support in-memory / disk / object store backends.

**Comment by wangrunji0408** [comment]
Isn't that only the disk backend depends on file operations?

**Comment by skyzh** [comment]
Yes. But even disk backend has two modes -- pure in-memory mode (for testing, where files are stored in a hash map), and real on-disk mode.

**Comment by skyzh** [comment]
If we can abstract all disk operations to use a trait like ObjectStore, we can prevent unwanted writes to disk.

**Comment by wangrunji0408** [comment]
I think we should use the real on-disk mode in testing, to make sure we are correctly using the system fs API. Data path can be redirected to ramfs/tmpfs to speed up.

**Comment by wangrunji0408** [comment]
Later we can also introduce the simulation testing, where all fs API will be mocked to a in-memory simulator.

**Comment by skyzh** [comment]
I switch to pure in-memory mode because tmpfs is slow 🤣 fsync and manifest write will add 1s latency to every disk test case, which looks weird.

Maybe I can try switch off fsync for all writes (even manifest), and maybe things will work.

**Comment by wangrunji0408** [comment]
🤣 okay let's stay in memory mode for efficiency.
